### ISSUE TITLE
BelongsToThrough relationship needed

### ISSUE THREAD

**Issue by upngo** [laravel/framework#6161]
Originally brought up and dismissed in #3052. (see for more examples)

The HasManyThrough relationship can not be used in both directions.

The simplest analogy is
person a knows person b's phone number
person b know's person c's phone number.

Person c can get person a's phone number via HasManyThrough.

Hasmanythrough does NOT work for getting from person a to person c as assumed (the reason #3052 was closed.

Related is #3786 where @taylorotwell commented  "How can hasManyThrough NOT work both ways. If it works one way, it by definition must work the "other way".
HasManyThrough cannot work this way because C does not point to B (foreign key), and B does not point to A (fk).

Currently two relationships are needed to represent this:
    A belongs to B
    B belongs to C.

**Comment by tpavlek** [comment]
Ran into this issue today, +1 from me.

**Comment by hotmeteor** [comment]
+1

**Comment by kalvisbuls** [comment]
+1

**Comment by gmhenderson** [comment]
+1

**Comment by votinux** [comment]
+1

**Comment by memtoko** [comment]
Guys, if you find your self need access both ways, don't use hasManyThrough relation. You should know how your data need accessed. It's wrong if you design your table like this but you need access user from permission.

````
users:
    id
   name
roles:
   id
   name
   user_id
permissions:
   id
   role_id
   name
````
The solutions for them is create relations for each model. For example many to many, here the example:

````
user has many role
use has many permission
role has many user
role has many permission
permission has many role
permission has many user
````

The table for such as model structure is something like this

````
users:
   id
   name
roles:
  id
  name
roles_users
  id
  role_id
  user_id
permissions:
  id
  name
permissions_roles
  id
  role_id
  permission_id
permissions_users
  id
  user_id
  permission_id
````    
After that setup you will get access both ways. From permission you can accesess related user and role, and is also possible from user to access related role and permission.

**Comment by tpavlek** [comment]
@memtoko this is such an absurd post. While I appreciate that you're attempting to "solve" this - what you're proposing is not a solution. Designing databases that are properly normalized is not "wrong".

A user doesn't have permissions on a 1-to-1 in most cases, they have roles, and the roles assign the permissions. If I change the permissions for a role, all the users in that role should be affected. And I should *not* have to modify n rows of the table where n is the number of users. Because that's what you're suggesting, I'd have to change every row in the permissions_users table that was affected by the change, where if I model the data properly, as everyone in this thread is suggesting, I change a single row in role_permissions.

In short, this feature is simply missing. It doesn't mean the framework is horrible, it doesn't mean that Taylor doesn't understand databases, just simply the inverse of HasManyThrough has not been coded yet. And we would like it to be, at the earliest convenience.

**Comment by memtoko** [comment]
@tpavlek That's the problem of your request, You need access in **not most case**. Because you always need to solve "here a user, are this user has permission to doing this actions?" If you want users via permisssion, it's already available via roles property in permission if you define the structure of has many through relations. You don't need belongThrough, as you always access them via the intermedite relations.

Here the example taken from Laravel docs:
````
countries
    id - integer
    name - string

users
    id - integer
    country_id - integer
    name - string

posts
    id - integer
    user_id - integer
    title - string
````
It's make sense if you want to try to display posts in specifix country, so you define hasManyThrough relations. But if you need **country** when display **a post**, you already have it via property user. You don't need to define method country in a post, because that country is a data coming from the user.

**Comment by memtoko** [comment]
But if your models has connections many-to-many with another models through third models, and that is work in both ways as your request, then the tables may be something like this, you should add row a_id, b_id in intermediate table.

````
a
  id
  name
b
  id
  name
c
  id
  b_id
  c_id
  name
  etc
````
The real example coming from Rails, here their docs: http://guides.rubyonrails.org/association_basics.html#the-has-many-through-association

**Comment by hotmeteor** [comment]
Thanks, but let's just move on with the topic of the ticket.

**Comment by memtoko** [comment]
@hotmeteor For me it's clear by the example. You don't need belongToThrough. If the relations is many-to-many that proceeding through a third model, then you can use same structure in link I provided. But if it just one-to-many, then you can think it like post-country and user

**Comment by myjeux** [comment]
+1

**Comment by GrahamCampbell** [comment]
Please don't +1 issues. It's not useful to use. It's already clear this feature is liked by lots of people. Any kind of helpful input is much more useful.

**Comment by adamwathan** [comment]
I'm having a hard time seeing an actual use-case for this. There's no reason to ever need the actual relationship object (there's not a hell of a lot you can do to a belongsTo relationship), so why can't you just walk the chain to get the object you want?

For example, I would've implemented @joshuajabbour's example this way:

```php
class Store extends Eloquent
{
    public function categories()
    {
        return $this->hasMany('Category');
    }

    public function items()
    {
        return $this->hasManyThrough('Item', 'Category');
    }
}

class Category extends Eloquent
{
    public function items()
    {
        return $this->hasMany('Item');
    }

    public function store()
    {
        return $this->belongsTo('Store');
    }
}

class Item extends Eloquent
{
    public function category()
    {
        return $this->belongsTo('Category');
    }

    public function getStoreAttribute()
    {
        return $this->category->store;
    }
}
```

Is there any situation where you would actually want to call `$item->store()->someMethodOnTheRelationshipObjectItself()` and not just retrieve the actual store object?

**Comment by tpavlek** [comment]
@adamwathan because I supported this issue I'll give a potential use case that I might see - and it's an optimization concern. 

Assume I have a Collection of `Item`s of arbitrary size `n` (arbitrary is important as this can potentially be very large). What I want is to get a list of all distinct `Store`s offering any of these items (assuming you're building a muti-tenant marketplace as I've done in the past then this is a *very* common use-case). So what are we left to do? If we iterate through the Collection of `Item`s calling `->getStoreAttribute` on each one, it's going to be a SQL call per.

Executing SQL statements in a loop of arbitrary size is bad, we all can agree on that. However, if I were able to get the *relationship* object, I could easily do a left fold of my `Item` collection, reducing it down to a single `Builder` containing the information I've retrieved from each relationship, then perform some distinct magic and suddenly I've reduced `n` SQL calls down to a constant `1`.

It just feels weird to not have this method available, as the other relationship methods have an inverse, but not `hasManyThrough`.

**Comment by Gufran** [comment]
I was just wondering why this issue is attracting so many people. If you are going to need a `BelongsToThrough` relationship that means you already have a `BelongsTo` relation which you want to chain further, _but in a single method call._

```php
class State extends Eloquent {
    public function country()
    {
        return $this->belongsTo(Country::class);
    }
}

class City extends Eloquent {
    public function state()
    {
        return $this->belongsTo(State::class);
    }

    public function country()
    {
        return $this->state->country();
    }
}
```

There you have it.

```php
City::find(3)->country(); // Gives you a BelongsTo relation
```

This will make same database calls as `belongsToThrough` is supposed to be.
Did I miss something very obvious ? I cant really see why you need a whole new Relationship.

\**Just trying to help. dont yell at me please.*\*

**Comment by upngo** [comment]
@Gufran The first comment on this said something similar to which the reply was, what if state() returns null? Then you're calling a relationship on a null object which will throw an exception as opposed to returning null as an empty relationship should. Also, the function country() is not a relationship method as we would like and thus we cannot use features like eager loading

**Comment by upngo** [comment]
@adamwathan It's not about retrieving the relationship object. The fact that it would be a relationship object does mean that we can use eager loading though which is useful. If we traverse the tree then everywhere we want a->b->c we will have to cater for b potentially being null either in each model we require a belongsToThrough or each time we want to use the relationship

**Comment by ablunier** [comment]
+1

**Comment by tomsowerby** [comment]
+1
One of the benefits of the "through" methods is that the connecting model doesn't need to be defined. Making tree traversal difficult.

**Comment by JcRicaro** [comment]
```
$subCategory = $this->belongsTo('TransactionSubCategory', 'sub_category_id');
        
return $subCategory->getResults()->belongsTo('TransactionCategory');
```

This is what I currently use guys, but sometimes I get no results for __$subCategory__, does anyone know how to purposely send a blank relationship object?

**Comment by phroggyy** [comment]
I'm considering delving into the Laravel framework source to resolve this. However, I would like to hear some opinions with regards to implementation.

The current `hasManyThrough` can work because the intermediary table has a reference to the third (or first, depending on how you see it) table.

Let's assume we have a `user`, who has `posts`, which have `comments` attached to them. Our tables could look something like this:

~~~
`users`
    id
    email
    password

`posts`
    user_id
    title
    body

`comments`
    post_id
    body
    [user_id]
~~~

So, it is obvious that we can establish a `hasManyThrough` relationship here; we simply find all posts with a given user id, and then get the comments that hold the associated `post_id`.

However, if we want to establish a `belongsToThrough` relationship, it's slightly more difficult. We can't do it the way `hasManyThrough` works. However, we could do the following:

  1. Assume that our `App\Post` model has a `user` method.
  2. We create a function that works like `hasManyThrough('App\User', 'App\Post', ['customNamedBelongsToMethodOnPostModel'], ['custom_column_by_which_comment_is_related_to_post'])`
  3. This function will call `user()` on the `Post` model and return it (hydrated of course)
  4. If we can't find a method `user()`, we can, instead of throwing an error, assume that it's a relationship of `$this->belongsTo('App\User')`

**Putting it in a general form**

~~~
// Let's assume we have three models – 'current', 'intermediate', and 'related'
// The general form for our function is then (in pseudo code)

public function hasManyThrough($related, $intermediate) {
    $intermediate = new $intermediate;
    $relatedClassName = (new ReflectionClass($related))->name
    if (method_exists($intermediate, $relatedName)) {
        return $intermediate->$relatedClassName
    }
    return $intermediate->belongsTo($related);
}
~~~

Something like that should work. I'll try to implement this, and also include the overrides

**Comment by arkeidolon** [comment]
+

**Comment by elliotlings** [comment]
+1

**Comment by GrahamCampbell** [comment]
Please stop +1ing issues. We already know people want this. A +1 is just annoying and isn't helping anything.

**Comment by webdevtasolutions** [comment]
+1 for @GrahamCampbell's cmnt ;)

**Comment by znck** [comment]
[belongsToThrough](https://github.com/znck/belongs-to-through) for Eloquent.

**Comment by yasinkocak** [comment]
+1

**Comment by robmarston** [comment]
+1

**Comment by GrahamCampbell** [comment]
> Please don't +1 issues. It's not useful to use. It's already clear this feature is liked by lots of people. Any kind of helpful input is much more useful.

**Comment by robmarston** [comment]
@GrahamCampbell apologies. I assumed the more +1's the higher priority this feature would be when considering which requested features to develop next -- is there some other way you all determine popularity or do you simply give them all equal consideration? As always, thank you for contributing :)

**Comment by intositeme** [comment]
I managed to get it to work how i wanted.  
**Country -> User -> Post**

To get the Country of a post, in the Post Model i added a mutator method so that getting the country need not be a function  

    public function getCountyAttribute($value)
    {
        return $this->user->country;
    }

So now i can just call  

    $post->country;   

To get past the n+1 (i see more than n+1, it depends on how many users you have and that count is added to n+1+count ), i can eager load

    $post = Post::with('user.country')->get();

    $post = Post::with('user')->get(); //this results with n+1

**Comment by znck** [comment]
@taylorotwell 

```yml
# Sample table structure (assumed by BelongsToThrough relation)
countries:
  id: unsigned int, primary
  name: string

states:
  id: unsigned int, primary
  name: string
  country_id: unsigned int, foreign(countries)

cities:
  id: unsigned int, primary
  name: string
  state_id: unsigned int, foreign(states)
users:
  city_id: unsigned int, foreign(cities)
```

Using BelongsToThrough relation:

```php

class City {
  public function country() {
    return $this->belongsToThrough(Country::class, State::class);
  }
}

class User {
  public function country() {
    return $this->belongsToThrough(Country::class, [State::class, City::class]);
  }
}
```

**Comment by EliasZ** [comment]
:+1:

**Comment by rdev5** [comment]
Just wanted to add another use case (as opposed to contributing to all the pretty +1's which I would do myself :)

Without boring everyone with all the details of a specific to an application I'm working on, I would propose that given a purely abstract example of a "forest" where Forest has many Trees which has many Branches which has many Leaves, I would find valid utility (as it appears most others would here as well) in being able to find out which Tree this Leaf on the ground came from, as in `Leaf belongsTo Tree Through Branch`.

If `belongsTo` is a path to the parent of a child node, then `belongsToThrough` ought to be implemented as a simply path to the grand parent of a child node because what happens at grandmas stays at grandmas :P

**Comment by GrahamCampbell** [comment]
> I assumed the more +1's the higher priority this feature would be when considering which requested features to develop next

Proposal issues generally are not considered. They're only there for the community to implement.

**Comment by westhacker** [comment]
+1 for belongsToThrough
